README.md
<div align="center">    

# torch-influence

![GitHub release (latest SemVer)](https://img.shields.io/github/v/release/alstonlo/torch-influence)
[![Read the Docs](https://img.shields.io/readthedocs/torch-influence)](https://torch-influence.readthedocs.io/en/latest/)
[![License: Apache 2.0](https://img.shields.io/badge/License-Apache_2.0-blue.svg)](LICENSE.txt)

</div>

![](examples/dogfish_influences.png)

torch-influence is a PyTorch implementation of influence functions, a classical
technique from robust statistics that estimates the effect of removing a single training data point on a model’s
learned parameters. In their seminal paper _Understanding Black-box Predictions via Influence Functions_
([paper](https://arxiv.org/abs/1703.04730)),
Koh & Liang (2017) first co-opted influence functions to the domain of machine learning. Since then,
influence functions have been applied on a variety of machine learning tasks,
including explaining model predictions, dataset relabelling and reweighing,
data poisoning, increasing model fairness, and data augmentation.

This library aims to be simple and minimal. In addition, it fixes a few errors found in some of the existing
implementations of influence functions.

Note: We plan to include the JAX version of this repository soon.

______________________________________________________________________

## Installation

Pip from source:

```bash
git clone https://github.com/alstonlo/torch-influence
 
cd torch_influence
pip install -e .   
 ```

______________________________________________________________________

## Quickstart

### Overview

In order to use torch-influence, the first step is to subclass its `BaseInfluenceModule` class and implement its
single abstract method `BaseInfluenceModule.inverse_hvp()`. This method computes inverse Hessian-vector products (iHVPs), 
which is an important but costly step in influence function computation. Conveniently, torch-influence provides three 
subclasses out-of-the-box:


<div align="center">
 
| Subclass  | Method of iHVP computation |
| ------------- | ------------- |
| `AutogradInfluenceModule`  | Direct computation and inversion of the Hessian with `torch.autograd`  |
| `CGInfluenceModule`  | Truncated Conjugate Gradients (Martens et al., 2010) ([paper](https://www.cs.toronto.edu/~jmartens/docs/Deep_HessianFree.pdf)) |
| `LiSSAInfluenceModule`  | Linear time Stochastic Second-Order Algorithm (Agarwal et al., 2016) ([paper](https://arxiv.org/abs/1602.03943)) |

</div>

The next step is to subclass `BaseObjective` and implement its four abstract methods. 
The `BaseObjective` class serves as an adapter that holds project-specific information about how 
training and test losses are computed. 
All of `BaseInfluenceModule` and its three subclasses require an implementation of `BaseObjective` to be passed through its constructor.
The following is a sample subclass for an $L_2$-regularized classification model:


```python
import torch
import torch.nn.functional as F
from torch_influence import BaseObjective


class MyObjective(BaseObjective):

    def train_outputs(self, model, batch):
        return model(batch[0])

    def train_loss_on_outputs(self, outputs, batch):
        return F.cross_entropy(outputs, batch[1])  # mean reduction required

    def train_regularization(self, params):
        return 0.01 * torch.square(params.norm())

    # training loss by default taken to be 
    # train_loss_on_outputs + train_regularization

    def test_loss(self, model, params, batch):
        return F.cross_entropy(model(batch[0]), batch[1])  # no regularization in test loss
```

Finally, all that is left is to piece everything together. 
After instantiating a subclass of `BaseInfluenceModule`, 
influence scores can then be computed through the `BaseInfluenceModule.influences()` method.
For example:

```python 
from torch_influence import AutogradInfluenceModule
   

module = AutogradInfluenceModule(
    model=model,
    objective=MyObjective(),  
    train_loader=train_loader,
    test_loader=test_loader,
    device=device,
    damp=0.001
)

# influence scores of training points 1, 2, and 3 on test point 0
scores = module.influences([1, 2, 3], [0])
```


For more details, we refer users to the [API Reference](https://torch-influence.readthedocs.io/en/latest/).



### Dogfish 

The `examples/` directory contains a more complete example, which finetunes the topmost
layer of a pretrained Inceptionv3 network on the Dogfish dataset (Koh & Liang, 2017). Then, it 
uses influence functions to find the most helpful and harmful training images,
with respect to a couple of test images. To run the example, please download and extract
the Dogfish dataset ([CodaLab](https://worksheets.codalab.org/bundles/0x550cd344825049bdbb865b887381823c))
into the `examples/` folder and execute the following:


```bash
# install dependencies
pip install -e .[dev]  

cd examples/

# train model and analyze influence scores
python analyze_dogfish.py  
```  

______________________________________________________________________

## Contributors

- [Alston Lo](https://github.com/alstonlo)
- [Juhan Bae](https://www.juhanbae.com/)



CHANGELOG.md
# Changelog

## 0.1.0 (2022-07-01)


### Features

* initial release ([#1](https://github.com/alstonlo/torch-influence/issues/1)) ([3b021fb](https://github.com/alstonlo/torch-influence/commit/3b021fb0e970460fdba18cbb13243ed62c8ffb32))


tests/test_base.py
import pytest
import torch
import torch.nn.functional as F
from torch import nn
from torch.utils import data

from torch_influence import AutogradInfluenceModule, BaseObjective, CGInfluenceModule


# pylint: disable=redefined-outer-name,protected-access

DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")


def make_mock_module(module_cls, batch_size, dtype):
    torch.manual_seed(100)

    mlp = nn.Sequential(
        nn.Linear(10, 32),
        nn.BatchNorm1d(32),
        nn.Tanh(),
        nn.Dropout(p=0.5),
        nn.Linear(32, 32),
        nn.BatchNorm1d(32),
        nn.Tanh(),
        nn.Dropout(p=0.5),
        nn.Linear(32, 1),
    ).to(dtype)

    class MockObjective(BaseObjective):

        def train_outputs(self, model, batch):
            return model(batch[0])

        def train_loss_on_outputs(self, outputs, batch):
            return F.mse_loss(outputs, batch[1])

        def train_regularization(self, params):
            return params.norm() ** 2.0

        def test_loss(self, model, params, batch):
            return F.mse_loss(model(batch[0]), batch[1])

    dataset = data.TensorDataset(torch.randn((50, 10), dtype=dtype), torch.randn((50, 1), dtype=dtype))
    train_loader = data.DataLoader(dataset, batch_size=batch_size)
    test_loader = data.DataLoader(dataset, batch_size=batch_size)

    return module_cls(
        model=mlp,
        objective=MockObjective(),
        train_loader=train_loader,
        test_loader=test_loader,
        device=DEVICE,
        damp=0.01,
    )


@pytest.fixture()
def module_f32_b7():
    return make_mock_module(CGInfluenceModule, batch_size=7, dtype=torch.float32)


@pytest.fixture()
def module_f32_bf():
    return make_mock_module(CGInfluenceModule, batch_size=50, dtype=torch.float32)


@pytest.fixture()
def module_f64():
    return make_mock_module(AutogradInfluenceModule, batch_size=7, dtype=torch.float64)


def test_param_reshaping(module_f32_b7):
    params = module_f32_b7._model_params(with_names=False)

    flat_params = module_f32_b7._flatten_params_like(params)
    assert flat_params.shape == (1569,)

    zeros = torch.zeros_like(flat_params)
    zeros = module_f32_b7._reshape_like_params(zeros)
    assert isinstance(zeros, tuple) and len(zeros) == 10


def test_make_functional(module_f32_b7):
    parmas1 = module_f32_b7._model_params(with_names=False)
    params2 = module_f32_b7._model_make_functional()

    assert len(parmas1) == len(params2)
    assert all(torch.equal(*ps) for ps in zip(parmas1, params2))

    with torch.no_grad():
        module_f32_b7._model_reinsert_params(params2, register=True)
    params3 = module_f32_b7._model_params(with_names=False)

    assert len(parmas1) == len(params3)
    assert all(torch.equal(*ps) for ps in zip(parmas1, params3))


def test_precision(module_f64):
    module = module_f64
    assert module.train_loss_grad([1, 2]).dtype == torch.float64
    assert module.test_loss_grad([1, 20]).dtype == torch.float64

    vec = torch.randn(1569, device=DEVICE, dtype=torch.float64)
    assert module.inverse_hvp(vec).dtype == torch.float64


def test_index_error(module_f32_b7):
    with pytest.raises(IndexError):
        module_f32_b7.train_loss_grad([-1])
    with pytest.raises(IndexError):
        module_f32_b7.train_loss_grad([50])

    with pytest.raises(ValueError):
        module_f32_b7.stest([2, 2, 3, 4, 5])


@pytest.mark.parametrize(
    "loader_kwargs,batch_sizes",
    [
        ({}, [7] * 7 + [1]),
        ({"batch_size": 10}, [10] * 5),
        ({"sample_n_batches": 5}, [7] * 5),
        ({"batch_size": 1, "subset": [1, 2, 3]}, [1] * 3),
        ({"subset": list(range(12))}, [7, 5]),
    ],
)
def test_loader_wrapper(module_f32_b7, loader_kwargs, batch_sizes):
    for i, (batch, batch_size) in enumerate(module_f32_b7._loader_wrapper(train=True, **loader_kwargs)):
        assert batch_size == batch_sizes[i]
        assert batch_size == batch[0].shape[0]


def test_loss_grad_at_batch(module_f32_b7, module_f32_bf):
    g1 = module_f32_b7.train_loss_grad(list(range(20)))
    g2 = module_f32_bf.train_loss_grad(list(range(20)))
    assert torch.allclose(g1, g2)

    g31 = module_f32_b7.train_loss_grad(list(range(11)))
    g32 = module_f32_b7.train_loss_grad(list(range(11, 20)))
    g3 = (11 / 20) * g31 + (9 / 20) * g32
    assert torch.allclose(g2, g3)


tests/test_modules_lin.py
import numpy as np
import pytest
import torch
import torch.nn.functional as F
from scipy import stats
from sklearn import datasets
from torch import nn
from torch.utils import data

from torch_influence import AutogradInfluenceModule, BaseObjective, CGInfluenceModule, LiSSAInfluenceModule


# pylint: disable=too-few-public-methods

DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")


class LinRegObjective(BaseObjective):

    def __init__(self, l2):
        self.l2 = l2

    def train_outputs(self, model, batch):
        return model(batch[0])

    def train_loss_on_outputs(self, outputs, batch):
        return F.mse_loss(outputs, batch[1])

    def train_regularization(self, params):
        return 0.5 * self.l2 * torch.square(params.norm())

    def test_loss(self, model, params, batch):
        return F.mse_loss(model(batch[0]), batch[1])


class LinRegProblem:

    def __init__(self, module_cls, module_kwargs, batch_size=7, dtype=torch.float32, l2=1e-2):

        if module_cls is LiSSAInfluenceModule:
            lissa_defaults = {"repeat": 20, "depth": 200, "scale": 50}
            lissa_defaults.update(module_kwargs)
            module_kwargs = dict(lissa_defaults)

        self.module_cls = module_cls
        self.damp = module_kwargs["damp"]
        self.dtype = dtype
        self.l2 = l2

        self.n = 100
        self.d = 11

        X, Y = datasets.make_regression(
            n_samples=(self.n + 10),
            n_features=(self.d - 1),
            random_state=12345,
            n_informative=(self.d - 1),
            noise=1
        )

        Y = torch.tensor(Y, dtype=dtype).unsqueeze(1)
        X = torch.tensor(X, dtype=dtype)
        X = torch.cat([X, torch.ones_like(Y)], dim=-1)

        X_train, X_test = X[:self.n], X[self.n:]
        Y_train, Y_test = Y[:self.n], Y[self.n:]

        self.X_train = X_train
        self.Y_train = Y_train
        self.X_test = X_test
        self.Y_test = Y_test

        train_set = data.TensorDataset(X_train[:, :-1], Y_train)
        test_set = data.TensorDataset(X_test[:, :-1], Y_test)

        self.model = self.fitted_model()

        self.module = module_cls(
            model=self.model,
            objective=LinRegObjective(l2),
            train_loader=data.DataLoader(train_set, batch_size=batch_size),
            test_loader=data.DataLoader(test_set, batch_size=batch_size),
            device=DEVICE,
            **module_kwargs
        )

    def fitted_model(self, remove_idx=None):
        X, Y = self.X_train, self.Y_train
        if remove_idx is not None:
            X = np.delete(X, remove_idx, axis=0)
            Y = np.delete(Y, remove_idx, axis=0)
        n = X.shape[0]

        params = torch.inverse(X.T @ X + (n * self.l2 * 0.5) * torch.eye(self.d)) @ X.T @ Y

        pt_model = nn.Linear(self.d - 1, 1, bias=True)
        pt_model.weight = torch.nn.Parameter(params[:-1].T)
        pt_model.bias = torch.nn.Parameter(params[-1])
        pt_model = pt_model.to(device=DEVICE, dtype=self.dtype)
        return pt_model


@pytest.mark.parametrize(
    "idxs,train",
    [
        ([0], True),
        ([0], False),
        (list(range(8)), True),
        (list(range(8)), False)
    ],
)
def test_loss_grad_at_batch(idxs, train):
    problem = LinRegProblem(CGInfluenceModule, {"damp": 0.0})
    module = problem.module

    X = problem.X_train if train else problem.X_test
    Y = problem.Y_train if train else problem.Y_test
    X, Y = X[idxs].to(DEVICE), Y[idxs].to(DEVICE)
    w = torch.concat([problem.model.weight.T, problem.model.bias.unsqueeze(0)], dim=0).detach()

    analytic_grad = (2 * X.T @ (X @ w - Y) / len(idxs)) + ((problem.l2 if train else 0.0) * w)
    analytic_grad = analytic_grad.squeeze(1)

    grad_fn = module.train_loss_grad if train else module.test_loss_grad
    module_grad = grad_fn(idxs)

    assert torch.allclose(analytic_grad, module_grad, atol=1e-3, rtol=1e-3)


@pytest.mark.parametrize(
    "problem",
    [
        LinRegProblem(AutogradInfluenceModule, {"damp": 0.02, "check_eigvals": True}),
        LinRegProblem(CGInfluenceModule, {"damp": 0.02, "atol": 0.0}),
        LinRegProblem(CGInfluenceModule, {"damp": 0.02, "gnh": True, "atol": 0.0}),
        LinRegProblem(LiSSAInfluenceModule, {"damp": 0.02}, batch_size=64),
        LinRegProblem(LiSSAInfluenceModule, {"damp": 0.02, "gnh": True}, batch_size=64),
    ],
)
def test_inverse_hvp(problem):
    X = problem.X_train.to(DEVICE)
    hess = ((2 / problem.n) * X.T @ X) + (problem.damp + problem.l2) * torch.eye(problem.d, device=DEVICE)

    torch.manual_seed(12345)
    for _ in range(3):
        vec = torch.randn(problem.d, device=DEVICE, dtype=problem.dtype)
        analytic_ihvp = torch.inverse(hess) @ vec
        module_ihvp = problem.module.inverse_hvp(vec)

        tol = 0.015 if (problem.module_cls is LiSSAInfluenceModule) else 1e-3
        assert torch.allclose(analytic_ihvp, module_ihvp, atol=tol, rtol=tol)


@pytest.mark.parametrize(
    "problem,test_idxs",
    [
        (LinRegProblem(AutogradInfluenceModule, {"damp": 0.0}), list(range(3, 9))),
        (LinRegProblem(CGInfluenceModule, {"damp": 0.0, "atol": 0.0}), list(range(3, 9))),
        (LinRegProblem(LiSSAInfluenceModule, {"damp": 0.0}, batch_size=64), list(range(3, 9))),
    ],
)
def test_influence_scores(problem, test_idxs):
    module = problem.module

    test_batch = module.test_loader.dataset[test_idxs]
    test_batch = test_batch[0].to(DEVICE), test_batch[1].to(DEVICE)

    base_params = nn.utils.parameters_to_vector(module.model.parameters())
    base_loss = module.objective.test_loss(module.model, base_params, test_batch)

    train_idxs = list(range(problem.n))
    if_scores = module.influences(train_idxs=train_idxs, test_idxs=test_idxs).cpu()

    loo_scores = []
    for idx in train_idxs:
        loo_model = problem.fitted_model(remove_idx=idx)
        loo_params = nn.utils.parameters_to_vector(loo_model.parameters())
        score = module.objective.test_loss(loo_model, loo_params, test_batch) - base_loss
        loo_scores.append(score)
    loo_scores = torch.tensor(loo_scores)

    assert stats.pearsonr(loo_scores, if_scores)[0] > 0.9
    assert stats.spearmanr(loo_scores, if_scores)[0] > 0.9


tests/test_modules_log.py
import functools
import pathlib

import numpy as np
import pytest
import torch
import torch.nn.functional as F
import torchvision
import torchvision.transforms as T
from scipy import stats
from sklearn import linear_model
from torch import nn
from torch.utils import data

from torch_influence import BaseObjective, CGInfluenceModule, LiSSAInfluenceModule


# pylint: disable=too-few-public-methods

DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")


@functools.lru_cache()
def load_binary_mnist():
    data_dir = str(pathlib.Path(__file__).parents[1] / "data")
    digits = [1, 3]

    bin_mnist = []
    for train in (True, False):
        mnist = torchvision.datasets.MNIST(data_dir, train=train, download=True, transform=T.ToTensor())
        X, Y = mnist.data, mnist.targets
        idxs = torch.logical_or(Y == digits[0], Y == digits[1])

        Y = Y[idxs].unsqueeze(1)
        Y = torch.where(Y == digits[1], 1.0, 0.0)

        X = X[idxs].flatten(start_dim=-2) / 255.0
        X = torch.concat([X, torch.ones_like(Y)], dim=-1)

        bin_mnist.append((X, Y))
    return bin_mnist


class LogRegObjective(BaseObjective):

    def __init__(self, l2):
        self.l2 = l2

    def train_outputs(self, model, batch):
        return torch.sigmoid(model(batch[0]))

    def train_loss_on_outputs(self, outputs, batch):
        return F.binary_cross_entropy(outputs, batch[1])

    def train_regularization(self, params):
        return 0.5 * self.l2 * torch.square(params.norm())

    def test_loss(self, model, params, batch):
        return F.binary_cross_entropy_with_logits(model(batch[0]), batch[1])


class LogRegProblem:

    def __init__(self, module_cls, module_kwargs, batch_size=100, dtype=torch.float32, l2=2e-2):

        if module_cls is LiSSAInfluenceModule:
            lissa_defaults = {"repeat": 1, "depth": 5000, "scale": 10}
            lissa_defaults.update(module_kwargs)
            module_kwargs = dict(lissa_defaults)

        self.module_cls = module_cls
        self.damp = module_kwargs["damp"]
        self.dtype = dtype
        self.l2 = l2

        bin_mnist = tuple((X.to(dtype), Y.to(dtype)) for (X, Y) in load_binary_mnist())
        (self.X_train, self.Y_train), (self.X_test, self.Y_test) = bin_mnist

        self.n = self.X_train.shape[0]
        self.d = self.X_train.shape[1]

        train_set = data.TensorDataset(self.X_train[:, :-1], self.Y_train)
        test_set = data.TensorDataset(self.X_test[:, :-1], self.Y_test)

        self.model = self.fitted_model()

        self.module = module_cls(
            model=self.model,
            objective=LogRegObjective(l2),
            train_loader=data.DataLoader(train_set, batch_size=batch_size),
            test_loader=data.DataLoader(test_set, batch_size=batch_size),
            device=DEVICE,
            **module_kwargs
        )

    def fitted_model(self, remove_idx=None):
        X, Y = self.X_train, self.Y_train
        if remove_idx is not None:
            X = np.delete(X, remove_idx, axis=0)
            Y = np.delete(Y, remove_idx, axis=0)
        n = X.shape[0]

        C = 1 / (n * self.l2 * 0.5)
        sk_model = linear_model.LogisticRegression(C=C, solver="lbfgs", tol=1e-10, max_iter=1000, fit_intercept=False)
        sk_model.fit(X, Y.squeeze(-1))
        params = torch.tensor(sk_model.coef_, dtype=self.dtype)

        pt_model = nn.Linear(self.d - 1, 1, bias=True)
        pt_model.weight = torch.nn.Parameter(params[:, :-1])
        pt_model.bias = torch.nn.Parameter(params[:, -1])
        pt_model = pt_model.to(device=DEVICE, dtype=self.dtype)
        return pt_model


@pytest.mark.parametrize(
    "idxs,train",
    [
        ([0], True),
        ([0], False),
        (list(range(101)), True),
        (list(range(101)), False)
    ],
)
def test_loss_grad_at_batch(idxs, train):
    problem = LogRegProblem(CGInfluenceModule, {"damp": 0.0})
    module = problem.module

    X = problem.X_train if train else problem.X_test
    Y = problem.Y_train if train else problem.Y_test
    X, Y = X[idxs].to(DEVICE), Y[idxs].to(DEVICE)

    w = torch.concat([problem.model.weight.T, problem.model.bias.unsqueeze(0)], dim=0).detach()

    analytic_grad = (X.T @ (torch.sigmoid(X @ w) - Y) / len(idxs)) + ((problem.l2 if train else 0.0) * w)
    analytic_grad = analytic_grad.squeeze(1)

    grad_fn = module.train_loss_grad if train else module.test_loss_grad
    module_grad = grad_fn(idxs)

    assert torch.allclose(analytic_grad, module_grad, atol=1e-3, rtol=1e-3)


@pytest.mark.parametrize(
    "problem,test_idxs",
    [
        (LogRegProblem(CGInfluenceModule, {"damp": 0.0, "atol": 1e-8, "maxiter": 300}), [5, 100, 12]),
        (LogRegProblem(LiSSAInfluenceModule, {"damp": 0.0}), [5, 100, 12]),
    ],
)
def test_influence_scores(problem, test_idxs):
    module = problem.module

    test_batch = module.test_loader.dataset[test_idxs]
    test_batch = test_batch[0].to(DEVICE), test_batch[1].to(DEVICE)

    base_params = nn.utils.parameters_to_vector(module.model.parameters())
    base_loss = module.objective.test_loss(module.model, base_params, test_batch)

    train_idxs = list(range(problem.n))
    if_scores = module.influences(train_idxs=train_idxs, test_idxs=test_idxs).cpu()

    top_points = torch.argsort(if_scores).tolist()
    top_points = top_points[:20] + top_points[-20:]
    if_scores = if_scores[top_points]

    loo_scores = []
    for idx in top_points:
        loo_model = problem.fitted_model(remove_idx=idx)
        loo_params = nn.utils.parameters_to_vector(loo_model.parameters())
        score = module.objective.test_loss(loo_model, loo_params, test_batch) - base_loss
        loo_scores.append(score)
    loo_scores = torch.tensor(loo_scores)

    assert stats.pearsonr(loo_scores, if_scores)[0] > 0.9
    assert stats.spearmanr(loo_scores, if_scores)[0] > 0.9


examples/analyze_dogfish.py
import pathlib

import matplotlib.pyplot as plt
import matplotlib.transforms as mtrans
import numpy as np
import torch
import torch.nn.functional as F
import torchvision
from sklearn import linear_model
from torch import nn
from torch.utils import data
from tqdm import tqdm, trange

from torch_influence import BaseObjective, CGInfluenceModule


# Example-specific constants
BASE_DIR = pathlib.Path(__file__).parent
DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")

DOGFISH_RAW_PATH = BASE_DIR / "dataset_dog-fish_train-900_test-300.npz"
DOGFISH_EMB_PATH = BASE_DIR / "dataset_dog-fish_embeds.npz"

DOGFISH_RAW_MMAP = np.load(str(DOGFISH_RAW_PATH), mmap_mode="r")

L2_WEIGHT = 1e-4


def load_dogfish_examples(split, idxs):
    image = DOGFISH_RAW_MMAP[f"X_{split}"][idxs]
    label = DOGFISH_RAW_MMAP[f"Y_{split}"][idxs]
    return image, label


def load_inception_embeds():
    if DOGFISH_EMB_PATH.is_file():
        print(f"Using cached Inceptionv3 embeddings: {DOGFISH_EMB_PATH}")
        return np.load(str(DOGFISH_EMB_PATH))

    # load pretrained Inceptionv3
    model = torchvision.models.inception_v3(pretrained=True, transform_input=False)
    model.fc = nn.Identity()  # remove the last layer
    model.eval()
    model = model.to(DEVICE)

    embeds = {}

    for split in ("train", "test"):
        n = 2 * (900 if (split == "train") else 300)

        X, Y = [], []
        for idx in trange(0, n, 20, desc=f"Embedding Dogfish ({split})"):
            batch_idxs = [idx + offset for offset in range(20)]
            images, labels = load_dogfish_examples(split, batch_idxs)

            # getting Inceptionv3 embedding
            with torch.no_grad():
                images = torch.tensor(images, dtype=torch.float32, device=DEVICE)
                images = images.permute([0, 3, 1, 2])
                h = model(images).cpu().numpy()

            X.append(h)
            Y.append(labels)

        embeds[f"X_{split}"] = np.concatenate(X, axis=0).astype(np.float32)
        embeds[f"Y_{split}"] = np.concatenate(Y, axis=0).astype(np.float32)

    np.savez(str(DOGFISH_EMB_PATH), **embeds)
    return embeds


def fit_model(X, Y):
    C = 1 / (X.shape[0] * L2_WEIGHT)
    sk_clf = linear_model.LogisticRegression(C=C, tol=1e-8, max_iter=1000)
    sk_clf = sk_clf.fit(X.numpy(), Y.numpy())

    # recreate model in PyTorch
    fc = nn.Linear(2048, 1, bias=True)
    fc.weight = nn.Parameter(torch.tensor(sk_clf.coef_))
    fc.bias = nn.Parameter(torch.tensor(sk_clf.intercept_))

    pt_clf = nn.Sequential(
        fc,
        nn.Flatten(start_dim=-2),
        nn.Sigmoid()
    )

    pt_clf = pt_clf.to(device=DEVICE, dtype=torch.float32)
    return pt_clf


def captioned_image(model, embeds, split, idx, scores):
    x = embeds[f"X_{split}"][idx]
    y = embeds[f"Y_{split}"][idx].item()

    x = torch.tensor(x, device=DEVICE).unsqueeze(0)
    y_hat = model(x).round().item()

    # turn image into [0 .. 255] RGB image
    image, _ = load_dogfish_examples(split, idx)
    image = (image * 127.5) + 127.5
    image = image.clip(min=0.0, max=255.0)
    image = np.round(image).astype(np.uint8)

    # turn labels into human-readable strings
    class_names = ["dog", "fish"]
    label = class_names[int(y)]
    pred = class_names[int(y_hat)]

    if split == "test":
        score_caption = f"Test Loss: {scores[idx]:.5f}"
    else:
        score_caption = f"Influence: {scores[idx]:+.5f}"
    label_caption = f"Pred: {pred}, Label: {label}"
    return image, label_caption + "\n" + score_caption


def main():
    print(f"Using {DEVICE}")

    # ===========
    # Load model and data
    # ==========

    embeds = load_inception_embeds()

    X_train = torch.tensor(embeds["X_train"])
    Y_train = torch.tensor(embeds["Y_train"])
    train_set = data.TensorDataset(X_train, Y_train)

    X_test = torch.tensor(embeds["X_test"])
    Y_test = torch.tensor(embeds["Y_test"])
    test_set = data.TensorDataset(X_test, Y_test)

    clf = fit_model(X_train, Y_train)

    # ===========
    # Get indices of top 5 test images with highest test loss
    # ===========

    test_losses = F.binary_cross_entropy(clf(X_test.to(DEVICE)), Y_test.to(DEVICE), reduction="none")
    test_idxs = torch.argsort(test_losses, dim=0, descending=True)[:5]
    test_idxs = test_idxs.tolist()

    test_images = [captioned_image(clf, embeds, "test", idx, test_losses) for idx in test_idxs]

    # ===========
    # Initialize influence module using custom objective
    # ===========

    class BinClassObjective(BaseObjective):

        def train_outputs(self, model, batch):
            return model(batch[0])

        def train_loss_on_outputs(self, outputs, batch):
            return F.binary_cross_entropy(outputs, batch[1])

        def train_regularization(self, params):
            return L2_WEIGHT * torch.square(params.norm())

        def test_loss(self, model, params, batch):
            outputs = model(batch[0])
            return F.binary_cross_entropy(outputs, batch[1])

    module = CGInfluenceModule(
        model=clf,
        objective=BinClassObjective(),
        train_loader=data.DataLoader(train_set, batch_size=32),
        test_loader=data.DataLoader(test_set, batch_size=32),
        device=DEVICE,
        damp=0.001,
        atol=1e-8,
        maxiter=1000,
    )

    # ===========
    # For each test point:
    #   1. Get the influence scores for all training points
    #   2. Find the most helpful and harmful training points
    # The most helpful point is that which, if removed, most increases the loss at the
    # test point of interest (as predicted by the influence scores). Conversely, the most harmful
    # test point is that which most decreases the test loss if removed.
    # ===========

    helpful_images = []
    harmful_images = []

    all_train_idxs = list(range(X_train.shape[0]))
    for test_idx in tqdm(test_idxs, desc="Computing Influences"):
        # In practice, this can be further optimized since we are recomputing the
        # loss gradients over the entire training dataset. With enough space, these
        # gradients can be cached instead.
        influences = module.influences(train_idxs=all_train_idxs, test_idxs=[test_idx])

        helpful = captioned_image(clf, embeds, "train", influences.argmax(), influences)
        harmful = captioned_image(clf, embeds, "train", influences.argmin(), influences)

        helpful_images.append(helpful)
        harmful_images.append(harmful)

    image_grid = [test_images, helpful_images, harmful_images]

    # ===========
    # Plot image grid
    # ==========

    fig, axes = plt.subplots(nrows=3, ncols=len(image_grid[0]), sharex=True, sharey=True, figsize=(15, 8))
    plt.subplots_adjust(wspace=0.5, hspace=0.75)

    # plot images
    for row in zip(image_grid, axes):
        for ((image, caption), ax) in zip(*row):
            ax.set_title(caption, size="medium")
            ax.set(aspect="equal")
            ax.imshow(image)
            ax.set_xticks([])
            ax.set_yticks([])

    fig.tight_layout()

    # draw separating lines
    # Reference: https://stackoverflow.com/questions/26084231/draw-a-separator-or-lines-between-subplots
    def get_bbox(ax_):
        r = fig.canvas.get_renderer()
        return ax_.get_tightbbox(r).transformed(fig.transFigure.inverted())

    bboxes = np.array(list(map(get_bbox, axes.flat)), mtrans.Bbox).reshape(axes.shape)
    xmax = np.array(list(map(lambda b: b.x1, bboxes.flat))).reshape(axes.shape).max(axis=0)
    xmin = np.array(list(map(lambda b: b.x0, bboxes.flat))).reshape(axes.shape).min(axis=0)
    xs = np.c_[xmax[1:], xmin[:-1]].mean(axis=1)

    for x in xs:
        line = plt.Line2D([x, x], [0.03, 0.97], transform=fig.transFigure, color="black", linewidth=1.2)
        fig.add_artist(line)

    # write row labels
    row_labels = ["Test Image", "Most Helpful", "Most Harmful"]
    for ax, label in zip(axes[:, 0], row_labels):
        ax.set_ylabel(label, rotation=90, size="x-large", fontweight="bold")

    fig.savefig("dogfish_influences.png", dpi=300)


if __name__ == "__main__":
    main()
